src/fastpath/fastpath.c

```diff
@@ -902,3 +902,240 @@ void NORETURN fastpath_vm_fault(vm_fault_type_t type)
+        /* Dequeue dest from the notification queue */
+        ntfn_queue_dequeue_fp(dest, ntfn_ptr);
+    }
+
+    /* Wake up the signalled thread and tranfer badge */
+    setRegister(dest, badgeRegister, badge);
+    thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+
+    /* Donate SC if necessary. The checks for this were already done before
+     * the point of no return */
+    maybeDonateSchedContext_fp(dest, sc);
+
+    /* Left this in the same form as the slowpath. Not sure if optimal */
+    if (sc_sporadic(dest->tcbSchedContext)) {
+        assert(dest->tcbSchedContext != NODE_STATE(ksCurSC));
+        if (dest->tcbSchedContext != NODE_STATE(ksCurSC)) {
+            refill_unblock_check(dest->tcbSchedContext);
+        }
+    }
+
+    if (!schedulable) {
+        maskInterrupt(true, irq);
+        ackInterrupt(irq);
+        restore_user_context();
+        UNREACHABLE();
+    }
+
+    if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority) {
+
+        /* Get destination thread.*/
+        newVTable = TCB_PTR_CTE_PTR(dest, tcbVTable)->cap;
+
+        /* Get vspace root. */
+        cap_pd = cap_vtable_cap_get_vspace_root_fp(newVTable);
+
+        /* Ensure that the destination has a valid VTable. */
+        if (unlikely(!isValidVTableRoot_fp(newVTable))) {
+            slowpath_irq();
+        }
+
+        /* Need to test that the ASID is still valid */
+        asid_t asid = cap_vspace_cap_get_capVSMappedASID(newVTable);
+        asid_map_t asid_map = findMapForASID(asid);
+        if (unlikely(asid_map_get_type(asid_map) != asid_map_asid_map_vspace ||
+                    VSPACE_PTR(asid_map_asid_map_vspace_get_vspace_root(asid_map)) != cap_pd)) {
+            slowpath_irq();
+        }
+#ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
+        /* Ensure the vmid is valid. */
+        if (unlikely(!asid_map_asid_map_vspace_get_stored_vmid_valid(asid_map))) {
+            slowpath_irq();
+        }
+        /* vmids are the tags used instead of hw_asids in hyp mode */
+        stored_hw_asid.words[0] = asid_map_asid_map_vspace_get_stored_hw_vmid(asid_map);
+#else
+        stored_hw_asid.words[0] = asid;
+#endif
+
+        /* we're going to switch threads - charge the current thread */
+        ticks_t prev = getNextInterrupt();
+        /* charge the previous thread */
+        if (checkBudget()) {
+            commitTime();
+        }
+
+        if (isSchedulable(NODE_STATE(ksCurThread))) {
+            SCHED_ENQUEUE_CURRENT_TCB;
+        }
+        switchToThread_fp(dest, cap_pd, stored_hw_asid);
+        /* only set the next irq if we really have to */
+        ticks_t next = getNextInterrupt();
+        if (next < prev) {
+            setDeadline(next - getTimerPrecision());
+        }
+        /* update sc */
+        NODE_STATE(ksCurSC) = NODE_STATE(ksCurThread)->tcbSchedContext;
+    } else if (NODE_STATE(ksCurThread)->tcbPriority > dest->tcbPriority) {
+        tcbSchedEnqueue(dest);
+    } else {
```

Comment:
Make it clearer what you're doing here. If thread is of lower prio than current thread, enqueue. If thread is of equal prio, append. Maybe just a comment in the else would suffice.